File: src/creationalPatterns/singleton/singleton.kt
object Singleton {
    init {
        println("Singleton instance created")
    }

    fun showMessage() {
        println("Hello from Singleton!")
    }
}

fun main() {
    // Access the Singleton instance
    val singletonInstance1 = Singleton
    val singletonInstance2 = Singleton

    // The Singleton instance is created only once, so these instances are the same
    println(singletonInstance1 === singletonInstance2) // Should print true

    // Call a method on the Singleton
    singletonInstance1.showMessage() // Should print "Hello from Singleton!"
    singletonInstance2.showMessage() // Should print "Hello from Singleton!"
}

File: src/creationalPatterns/singleton/singleton.ts
/**
 * The Singleton class defines the `getInstance` method that lets clients access
 * the unique singleton instance.
 */
class Singleton {
  private static instance: Singleton | null = null;

  /**
   * The Singleton's constructor should always be private to prevent direct
   * construction calls with the `new` operator.
   */
  private constructor() {}

  /**
   * The static method that controls the access to the singleton instance.
   *
   * This implementation let you subclass the Singleton class while keeping
   * just one instance of each subclass around.
   */
  public static getInstance(): Singleton {
    if (!Singleton.instance) {
      Singleton.instance = new Singleton();
    }

    return Singleton.instance;
  }

  /**
   * Finally, any singleton should define some business logic, which can be
   * executed on its instance.
   */
  public someBusinessLogic() {
    console.log('Ejecutando un método en la instancia Singleton.');
  }
}

/**
 * The client code.
 */
function clientCode() {
  const instance1 = Singleton.getInstance();
  const instance2 = Singleton.getInstance();

  if (instance1 === instance2) {
    console.log('Singleton works, both variables contain the same instance.');
  } else {
    console.log('Singleton failed, variables contain different instances.');
  }
  instance1.someBusinessLogic();
}

clientCode();
